test/linalg.jl

```diff
@@ -62,3 +64,57 @@ end
         end
     end
 end
+
+###### Test jacobian matrix ######
+params(A::SparseMatrixCSC) = params(A.nzval)
+fitparams(A::SparseMatrixCSC, p) = SparseMatrixCSC(A.m, A.n, A.colptr, A.rowval, fitparams(A.nzval, p))
+params(v::DenseArray{<:Real}) = v
+fitparams(v::DenseArray{<:Real}, p) = p
+params(v::DenseArray{<:Complex{T}}) where T = collect(reinterpret(T, v))
+fitparams(v::DenseArray{<:T}, p::DenseArray{T2}) where {T<:Complex,T2} = collect(reinterpret(Complex{T2}, p))
+
+@i function loss(l, j, y::AbstractArray{<:Real}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    l += y[j]
+end
+@i function loss(l, j, y::AbstractArray{<:Complex}, A, x)
+    imul!(y, A, x, 1.0, 1.0)
+    if j%2 == 1
+        l += y[div(j-1,2)+1].re
+    else
+        l += y[div(j-1,2)+1].im
+    end
+end
+
+function nilang_mul_jacobian(A, x::AbstractArray)
+    px, pA = params(x), params(A)
+    J = zeros(T, length(px), length(px) + length(pA))
+    for j=1:length(px)
+        _, _, _, _, gA, gx = NiLang.AD.Grad(loss)(Val(1), 0.0, j, zero(x), A, x)
+        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
+    end
+    return J
```

Comment:
This is reverse mode AD...

Reply:
Sorry, I do not understand your question, can not please say it in a different way?

Reply:
In this function you calculate the $\frac{\partial{f}}{\partial{pAx}}$. However, maybe only $\frac{\partial{f}}{\partial{pA}}$ or $\frac{\partial{f}}{\partial{px}}$ is needed. Should we implement for such special cases so as to avoid unnecessary computation?

Reply:
I do not think so. It does not save much computational effort but will introduce too much complexity. Will not be considered in current stage. Keep the code simple is important.